---
title: "DNA Sequencing Analysis"
output: html_document
---

Here is a general reference on sequence analysis in R: https://www.bioconductor.org/help/course-materials/2013/useR2013/Bioconductor-tutorial.pdf

Use the analysis of epPCR data described in this paper: https://www-sciencedirect-com.proxy.library.upenn.edu/science/article/pii/S109671761830123X

Here is the link to their data files:
https://data.mendeley.com/datasets/wkd5wp9zwk/1

First, they used dada2 to preprocess their NGS sequencing reads:
https://benjjneb.github.io/dada2/tutorial.html

Then they worked up the data using a custom analysis just like we want, but they did it in Python! Here is the link to their python notebook (similar to a .Rmd):
https://data.mendeley.com/datasets/wkd5wp9zwk/1/files/5e4bed9a-f7d0-4d08-913b-58001cc9ea35

I want you to recreate their same analysis (minus a few of the downstream bits, we can discuss) but in R, so you will need to translate between these languages (I think this is actually a great way for you to learn).

For now, use their data with 300bp reads as example data, it is similar enough to what I'm hoping to gather in the next few weeks.

Experiment: Use error prone DNA polymerase to amplify parts of CSF1R gene with random mutations.
Goals of Analysis: Obtain the mutation rate, and the uniformity of the mutation rate across a string of DNA read.

Good luck!

```{r setup, include=FALSE}
library(dada2)
library(tidyverse)
library(Biostrings) 
```


```{r}
path <- "/Users/grahampeet/Downloads/MiSeq_SOP/"
#I'm using just two files from the example data used by the dada tutorial, it's a lot faster to work with.

  fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
  fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
  sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
  plotQualityProfile(fnFs[1:2])
  plotQualityProfile(fnRs[1:2])

```

```{r}
path2 <- "/Users/grahampeet/Desktop/CSF1R_Seq"

filtFs <- file.path(path2, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path2, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

#Here we have a function that filters outlying reads and can truncate the region of the genome to exclude regions where the data has poor quality.
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs,
                     multithread = T)
head(out)
```

```{r}
learnErrors(filtFs, multithread=TRUE) %>%
  dada(filtFs, err=., multithread=TRUE) %>%
  plotErrors(nominalQ=TRUE)

learnErrors(filtRs, multithread=TRUE) %>%
  dada(filtRs, err=., multithread=TRUE) %>%
  plotErrors(nominalQ=TRUE)

errF <- learnErrors(filtFs, multithread = T)
errR <- learnErrors(filtRs, multithread = T)

InferenceFs <- dada(filtFs, err=errF, multithread=TRUE)
InferenceRs <- dada(filtRs, err=errR, multithread=TRUE)

InferenceFs[[1]]

#Combine paired reads here before proceeding
```


```{r}
InferenceFs %>%
  makeSequenceTable() %>% 
  t() %>%
  as_tibble(rownames = "Sequence") -> GP_testing

#use this function, I couldn't make your old code run without a stack overflow error
head(GP_testing)
```


```{r GP Functions}
#This function takes strings that have already been split into vectors
mutfinderGP <- function(test_seq, ref_seq) {
  if (length(test_seq) == length(ref_seq)) { #needs to be same length to analyze
    if (test_seq == ref_seq) { #check if WT
        return(test_seq) #code to return current line
      else {# find SNPs 
        return(map2(test_seq, ref_seq, list) %>%
                 imap(~ if((.[[1]] != .[[2]])){paste(.y, .[[2]], ">", .[[1]], sep="")}) %>%
                 unlist())
      } 
      else { 
        return("indel")
      }
    } 
  }
}


######This code only for testing
test_seq <- unlist(str_split("ATGG", pattern = ""))
ref_seq <- unlist(str_split("ATCA", pattern = ""))

map2(test_seq, ref_seq, list) %>%
  imap(~ if((.[[1]] != .[[2]])){paste(.y, .[[2]], ">", .[[1]], sep="")}) %>%
  unlist()

```


```{r}
#This code isn't tested, but should do the right thing.
df <- GP_testing
ref_seq <- character(read_csv("ref_seq_300.csv")) #or as fasta or whatever

df_freq <- tibble()

df_freq$muts <- map(df$Sequence, mutfinderGP(df$Sequence, ref_seq))
df_freq$num <- length(df$muts)

df <- cbind(df, df_freq)
```


